promql/test.go

```diff
@@ -427,6 +428,79 @@ func (t *Test) Run() error {
 	return nil
 }
 
+type atModifierTestCase struct {
+	expr     string
+	evalTime time.Time
+}
+
+func atModifierTestCases(exprStr string, evalTime time.Time) ([]atModifierTestCase, error) {
+	expr, err := parser.ParseExpr(exprStr)
+	if err != nil {
+		return nil, err
+	}
+	ts := timestamp.FromTime(evalTime)
+
+	containsNonStepInvariant := false
+	// Setting the @ timestamp for all selectors to be evalTime.
+	// If there is a subquery, then the selectors inside it don't get the @ timestamp.
+	// If any selector already has the @ timestamp set, then it is untouched.
+	parser.Inspect(expr, func(node parser.Node, path []parser.Node) error {
+		_, _, subqTs := subqueryTimes(path)
+		if subqTs != nil {
+			// There is a subquery with timestamp in the path,
+			// hence don't change any timestamp further.
+			return nil
+		}
+		switch n := node.(type) {
+		case *parser.VectorSelector:
+			if n.Timestamp == nil {
+				n.Timestamp = makeInt64Pointer(ts)
+			}
+
+		case *parser.MatrixSelector:
+			if vs := n.VectorSelector.(*parser.VectorSelector); vs.Timestamp == nil {
+				vs.Timestamp = makeInt64Pointer(ts)
+			}
+
+		case *parser.SubqueryExpr:
+			if n.Timestamp == nil {
+				n.Timestamp = makeInt64Pointer(ts)
+			}
+
+		case *parser.Call:
+			containsNonStepInvariant = containsNonStepInvariant || !IsFunctionStepInvariant(n)
+			if n.Func.Name == "timestamp" {
+				// If timestamp(..) has something other than vector selector inside,
+				// then it is not step invariant.
+				_, ok := n.Args[0].(*parser.VectorSelector)
+				containsNonStepInvariant = containsNonStepInvariant || !ok
+			}
+		}
+		return nil
+	})
+
+	if containsNonStepInvariant {
```

Comment:
I can return an error inside parser.Inspect avoiding inspecting the entire expression, but that just adds more lines. I would leave it like this since it's simpler currently and it's for tests.

Reply:
I mean that as soon as containsNonStepInvariant is set to true, you can return.

Reply:
Since it is set inside the `parser.Inspect` callback, it would only exit `parser.Inspect` and we will have to check here anyway :)